File: app/src/main/java/com/sdex/activityrunner/intent/dialog/SingleSelectionDialog.java
package com.sdex.activityrunner.intent.dialog;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import com.sdex.activityrunner.intent.dialog.source.SelectionDialogSource;
import com.sdex.activityrunner.util.ObjectsCompat;
import java.util.ArrayList;

public class SingleSelectionDialog extends DialogFragment {

  public static final String TAG = "SingleSelectionDialog";

  private static final String ARG_TYPE = "arg_type";
  private static final String ARG_SOURCE = "arg_source";
  private static final String ARG_INITIAL_POSITION = "arg_initial_position";

  private OnItemSelectedCallback callback;

  public static SingleSelectionDialog newInstance(int type, SelectionDialogSource source,
    int initialPosition) {
    Bundle args = new Bundle(3);
    args.putInt(ARG_TYPE, type);
    args.putParcelable(ARG_SOURCE, source);
    args.putInt(ARG_INITIAL_POSITION, initialPosition);
    SingleSelectionDialog fragment = new SingleSelectionDialog();
    fragment.setArguments(args);
    return fragment;
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    final int type;
    final SelectionDialogSource source;
    final int initialPosition;
    if (getArguments() != null) {
      type = getArguments().getInt(ARG_TYPE);
      source = getArguments().getParcelable(ARG_SOURCE);
      initialPosition = getArguments().getInt(ARG_INITIAL_POSITION);
    } else {
      throw new NullPointerException();
    }
    final ArrayList<String> list = ObjectsCompat.requireNonNull(source).getList();
    final AlertDialog.Builder builder =
      new AlertDialog.Builder(ObjectsCompat.requireNonNull(getActivity()));
    builder.setSingleChoiceItems(list.toArray(new String[list.size()]),
      initialPosition, (dialog, which) -> {
        callback.onItemSelected(type, which);
        dismiss();
      });
    builder.setTitle(type);
    return builder.create();
  }

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    try {
      callback = (OnItemSelectedCallback) context;
    } catch (ClassCastException e) {
      throw new ClassCastException(context.toString()
        + " must implement OnItemSelectedCallback");
    }
  }

  public interface OnItemSelectedCallback {

    void onItemSelected(int type, int position);
  }
}

File: app/src/main/java/com/sdex/activityrunner/intent/dialog/SingleSelectionDialog.kt
package com.sdex.activityrunner.intent.dialog

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.sdex.activityrunner.R
import com.sdex.activityrunner.commons.BaseDialogFragment
import com.sdex.activityrunner.intent.dialog.source.ActionSource
import com.sdex.activityrunner.intent.dialog.source.MimeTypeSource

class SingleSelectionDialog : BaseDialogFragment() {

    private lateinit var callback: OnItemSelectedCallback

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val type: Int = requireArguments().getInt(ARG_TYPE)
        val initialPosition: Int = requireArguments().getInt(ARG_INITIAL_POSITION)

        val source = when (type) {
            R.string.launch_param_action -> ActionSource()
            R.string.launch_param_mime_type -> MimeTypeSource()
            else -> throw IllegalStateException("Wrong type: $type")
        }

        val list = source.list
        val builder = MaterialAlertDialogBuilder(requireActivity())
        builder.setSingleChoiceItems(
            list.toTypedArray(),
            initialPosition
        ) { _, which ->
            callback.onItemSelected(type, which)
            dismiss()
        }
        builder.setTitle(type)
        return builder.create()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            callback = context as OnItemSelectedCallback
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement OnItemSelectedCallback")
        }
    }

    interface OnItemSelectedCallback {

        fun onItemSelected(type: Int, position: Int)
    }

    companion object {

        const val TAG = "SingleSelectionDialog"

        private const val ARG_TYPE = "arg_type"
        private const val ARG_INITIAL_POSITION = "arg_initial_position"

        fun newInstance(type: Int, initialPosition: Int): SingleSelectionDialog {
            val args = Bundle(2)
            args.putInt(ARG_TYPE, type)
            args.putInt(ARG_INITIAL_POSITION, initialPosition)
            val fragment = SingleSelectionDialog()
            fragment.arguments = args
            return fragment
        }
    }
}
